### PR TITLE
Enable CORS for with OTLP HTTP endpoint and add playground app

### LINKED ISSUE #4917: Configuration to allow CORS with OTLP HTTP
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

Someone writing a rich browser app, e.g. React, might want to send telemetry via OTLP HTTP to the dashboard. The dashboard supports OTLP HTTP, but standard browser security blocks post requests from the browser app to the dashboard because they're hosted on different ports.

### Describe the solution you'd like

Configuration to enable CORS for the dashboard's OTLP HTTP endpoints.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/aspire/issues/4917

This PR:

* Builds on top of https://github.com/dotnet/aspire/pull/5177. The previous PR allowed the dashboard to be configured to support CORS. This PR improves the host to automatically configure the dashboard with the allowed origins of all HTTP endpoints in the Aspire app.
* Adds a playground app that uses the `opentelemetry-js` SDK to send OTLP data to the dashboard.

Demo:
browser-open-telemetry

The eagle eyed might have noticed that the browser span started before the server span. It's caused by the browser and server clocks not being perfectly synced. They are off by about 10ms. As far as I can tell there isn't anything that we can do about it.

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5212)

### CODE REVIEW

File: src/Aspire.Hosting/Dashboard/DashboardLifecycleHook.cs [hunk 1]
```diff
@@ -160,6 +161,32 @@ private void ConfigureAspireDashboardResource(IResource dashboardResource)
...
+                        if (endpoint.Port != null)
+                        {
+                            corsOrigins.Add($"{endpoint.UriScheme}://localhost:{endpoint.Port}");
+                        }
+                        if (endpoint.TargetPort != null)
+                        {
+                            corsOrigins.Add($"{endpoint.UriScheme}://localhost:{endpoint.TargetPort}");
+                        }
+                    }
+                }
```
Comment (JamesNK) [inline src/Aspire.Hosting/Dashboard/DashboardLifecycleHook.cs:186]: @mitchdenny Is this the right way to get all of the resource endpoints? I noticed that sometimes the `EndpointAnnotation.Port` value was missing or wrong.

For example, one endpoint in TestShop says it has a port of 80, which is wrong. But it has an `EndpointAnnotation.AllocatedEndpoint` which looks like it reports the correct port.

**Comment by JamesNK** [comment]
cc @samsp-msft @aaronpowell @DamianEdwards

File: src/Aspire.Hosting/Dashboard/DashboardLifecycleHook.cs (hunk 1 above)
Comment (mitchdenny) [reply src/Aspire.Hosting/Dashboard/DashboardLifecycleHook.cs:186]: Yeah allocated endpoint is what you want.

**Comment by mitchdenny** [comment]
Just thinking whether we need a threat model for this if only just to verify that we've though that we aren't making anything less secure by setting CORS up for all the resource endpoints. I don't think we do because someone who has compromised a web UI on one of the resource endpoints probably has control of the container the UI was launched from, at which point CORS isn't relevant.

File: src/Aspire.Hosting/Dashboard/DashboardLifecycleHook.cs (hunk 1 above)
Comment (mitchdenny) [reply src/Aspire.Hosting/Dashboard/DashboardLifecycleHook.cs:186]: Yeah, if you want the actual ports as it was determined by DCP then you want the allocated endpoint. Docs can definitely be improved.

If the Port on the EndpointAnnotation is null you are basically saying that you want DCP to allocate a random one.

File: tests/Aspire.Hosting.Tests/Dashboard/DashboardResourceTests.cs [hunk 2]
```diff
@@ -263,6 +263,68 @@ public async Task DashboardResourceServiceUriIsSet()
...
+        var dashboard = Assert.Single(model.Resources.Where(r => r.Name == "aspire-dashboard"));
+
+        var config = await EnvironmentVariableEvaluator.GetEnvironmentVariablesAsync(dashboard, DistributedApplicationOperation.Run, app.Services);
+
+        Assert.Equal("http://localhost:8080,http://localhost:58080", config.Single(e => e.Key == DashboardConfigNames.DashboardOtlpCorsAllowedOriginsKeyName.EnvVarName).Value);
+        Assert.Equal("*", config.Single(e => e.Key == DashboardConfigNames.DashboardOtlpCorsAllowedHeadersKeyName.EnvVarName).Value);
+    }
+
+    [Fact]
+    public async Task DashboardResource_OtlpGrpcEndpoint_CorsEnvVarSet()
```
Comment (JamesNK) [inline tests/Aspire.Hosting.Tests/Dashboard/DashboardResourceTests.cs:298]:
```suggestion
    public async Task DashboardResource_OtlpGrpcEndpoint_CorsEnvVarNotSet()
```

File: src/Aspire.Hosting/Dashboard/DashboardLifecycleHook.cs [hunk 3]
```diff
@@ -160,6 +161,42 @@ private void ConfigureAspireDashboardResource(IResource dashboardResource)
...
+                    .ToList();
+
+                var corsOrigins = new HashSet<string>();
+                foreach (var endpoint in allResourceEndpoints)
+                {
+                    if (endpoint.UriScheme is "http" or "https")
+                    {
+                        // Prefer allocated endpoint over EndpointAnnotation.Port.
+                        var origin = endpoint.AllocatedEndpoint?.UriString;
+                        if (origin == null && endpoint.Port != null)
```
Comment (mitchdenny) [inline src/Aspire.Hosting/Dashboard/DashboardLifecycleHook.cs:179]: Not sure if this fallback logic is necessary. We should always have an allocated endpoint at this stage.

File: src/Aspire.Hosting/Dashboard/DashboardLifecycleHook.cs [hunk 4]
```diff
@@ -160,6 +161,42 @@ private void ConfigureAspireDashboardResource(IResource dashboardResource)
...
+                        {
+                            origin = $"{endpoint.UriScheme}://localhost:{endpoint.Port}";
```
Comment (mitchdenny) [inline src/Aspire.Hosting/Dashboard/DashboardLifecycleHook.cs:181]: `localhost` could have an extreme corner case here. Lets say someone spins up a headless browser in a container (perhaps for test automation). Then localhost will likely be the wrong hostname (platform dependent).

**Review by mitchdenny — APPROVED** [review]
LGTM. Added a few comments but nothing that should block.

File: src/Aspire.Dashboard/DashboardWebApplication.cs [hunk 5]
```diff
@@ -148,7 +148,39 @@ public Func<EndpointInfo> OtlpServiceHttpEndPointAccessor
...
+                // Default policy allows the dashboard's origins.
+                // This is added so CORS middleware doesn't report failure for dashboard browser requests that include an origin header.
+                options.AddDefaultPolicy(builder =>
+                {
+                    builder.WithOrigins(dashboardOptions.Frontend.GetEndpointUris().Select(uri => uri.OriginalString).ToArray());
+                    builder.AllowAnyHeader();
+                    builder.AllowAnyMethod();
+                });
+
+                options.AddPolicy("OtlpHttp", builder =>
```
Comment (drewnoakes) [inline src/Aspire.Dashboard/DashboardWebApplication.cs:164]: Seems like `OtlpHttp` is a constant that's referenced again below in `OtlpHttpEndpointsBuilder.cs`. Extract to a shared location?

File: src/Aspire.Hosting/Dashboard/DashboardLifecycleHook.cs [hunk 6]
```diff
@@ -160,6 +161,42 @@ private void ConfigureAspireDashboardResource(IResource dashboardResource)
...
                 context.EnvironmentVariables[DashboardConfigNames.DashboardOtlpHttpUrlName.EnvVarName] = otlpHttpEndpointUrl;
+
+                var model = context.ExecutionContext.ServiceProvider.GetRequiredService<DistributedApplicationModel>();
+                var allResourceEndpoints = model.Resources
+                    .Where(r => !string.Equals(r.Name, KnownResourceNames.AspireDashboard, StringComparisons.ResourceName))
+                    .SelectMany(r => r.Annotations)
+                    .OfType<EndpointAnnotation>()
+                    .ToList();
+
+                var corsOrigins = new HashSet<string>();
```
Comment (drewnoakes) [inline src/Aspire.Hosting/Dashboard/DashboardLifecycleHook.cs:172]: Explicit comparator?

**Review by drewnoakes — APPROVED** [review]

File: src/Aspire.Hosting/Dashboard/DashboardLifecycleHook.cs (hunk 4 above)
Comment (JamesNK) [reply src/Aspire.Hosting/Dashboard/DashboardLifecycleHook.cs]: Let's wait and see whether there is feedback about the automatic CORS policy not working. Can figure out improvements then.

**Pull request merged**